lib/galaxy/util/xml_macros.py

```diff
@@ -134,48 +164,60 @@ def _expand_macros(elements, macros, tokens):
             expand_el = element.find('.//expand')
             if expand_el is None:
                 break
-            _expand_macro(element, expand_el, macros, tokens)
+            if visited is None:
+                v = list()
+            else:
+                v = visited
+            _expand_macro(expand_el, macros, tokens, v)
 
 
-def _expand_macro(element, expand_el, macros, tokens):
+def _expand_macro(expand_el, macros, tokens, visited):
     macro_name = expand_el.get('macro')
     assert macro_name is not None, "Attempted to expand macro with no 'macro' attribute defined."
+
+    # check for cycles in the nested macro expansion
+    assert macro_name not in visited, f"Cycle in nested macros: already expanded {visited} can't expand '{macro_name}' again"
```

Comment:
Are you sure this must be a cycle ? Also list lookups can be expensive, maybe make visited a set or an ordered set?

Reply:
> Are you sure this must be a cycle ? 

Quite. Essentially this is a DFS of the expansions and if a previously expanded macro is expanded again we have a cycle. Note that visisted is kind of a stack (i.e. the macro names are removed at the end of the function.)

> Also list lookups can be expensive, maybe make visited a set or an ordered set?

Then I will go for ordered set, since otherwise the test can not check for the message (or I just test for the prefix `"Cycle in nested macros"`.